### ISSUE TITLE
URI SAN requirement for signing certs rationale.

### ISSUE THREAD

**Issue by banks** [spiffe/spiffe#67]
Apart from the other issue I filed about how the current test suite is out of date w.r.t. signing cert SANs (#65), can I ask what the rationale is for requiring that signing certificates all must be valid SVIDs right back to the root?

All else equal it makes sense to specify a new format that way, however in practice I know of organisations who would want to root trust in any new system from an existing organisational trust root which may or may not be capable of adding URI SANs to the root and intermediates.

Further, several third-party CAs we need to integrate with don't support generating roots with URI SANs which rules out both implementing SPIFFE and being compatible with them.

I'd be interested to know the rationale for making that a strict requirement in the spec.

As it stands it _seems_ to me like many existing orgs will either need to implement non-conforming validation that allows for external non-compliant trust anchors, or will need to just avoid SPIFFE and come up with essentially the same thing without using the "standard".

I note for example that Envoy doesn't appear to even perform SPIFFE validation (just regular x509 path validation via openssl) despite being the implementation used in Istio which as far as I understand is the main motivating implementation. I don't know if that's just a timing thing but it seems like vendors would be tempted to not implement that validation strictly to maintain compatibility.

Any info here and potentially in the spec text on why this is deemed necessary, or consideration of relaxing it to allow compatibility would be great!

Thanks for pushing this standard forward.

**Comment by evan2645** [comment]
> can I ask what the rationale is for requiring that signing certificates all must be valid SVIDs right back to the root?

In this case, only the immediate signing certificate is required to conform in this way - the restriction does not apply from the root down. For instance, given self-signed root A, and intermediate B which has issued leaf C, A is not required to be a valid SVID. We have explicitly avoided putting requirements on the full chain for the reasons you describe.

As an example, the SPIRE implementation has a CA plugin and an Upstream CA plugin. The former issues a CSR for an intermediate CA certificate to the latter, where the CSR it generates is an SVID for the configured trust domain, and the upstream signing authority may be existing non-SPIFFE PKI.

That being said, I think it would make sense to switch that MUST to a SHOULD. The SPIFFE validation logic (as it relates to "federated bundles") was previously set up in such a way that required the immediate signing cert to be an SVID, however that logic changed last year and this MUST was left behind.

**Comment by banks** [comment]
Interesting thanks for the response.

I was also perhaps misunderstanding the spec - I took it to mean that every cert in the chain back to and including the root must be an SVID which is why federating trust to an external CA as you describe seemed incompatible.

If that is not the intent even with the current wording then it alleviates a lot of the concerns. Perhaps an explicit statement of that would be a good amendment although I will go and re-read as it may be it ways that clearly already and I misinterpreted.
